crates/transpiler/src/passes/alap_schedule_analysis.rs

```diff
@@ -0,0 +1,175 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::TranspilerError;
+use hashbrown::HashMap;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType, Wire};
+use qiskit_circuit::operations::{OperationRef, StandardInstruction};
+use qiskit_circuit::{Clbit, Qubit};
+use rustworkx_core::petgraph::prelude::NodeIndex;
+
+#[pyfunction]
+/// Runs the ALAPSchedule analysis pass on dag.
+///
+/// Args:
+///     dag (DAGCircuit): DAG to schedule.
+///     clbit_write_latency (u64): The latency to write classical bits.
+///     node_durations (HashMap<usize, f64>): Mapping from node indices to operation durations.
+///
+/// Returns:
+///     PyDict: A dictionary mapping each DAGOpNode to its scheduled start time.
+///
+#[pyo3(name = "alap_schedule_analysis", signature= (dag, clbit_write_latency, node_durations))]
+pub fn run_alap_schedule_analysis(
+    py: Python,
+    dag: &DAGCircuit,
+    clbit_write_latency: u64,
+    node_durations: HashMap<usize, f64>,
+) -> PyResult<Py<PyDict>> {
```

Comment:
If we're trying to keep everything rust native, is there a way you could return a HashMap for this too? We want to be able to run this pass standalone (that meaning running the pass entirely from rust without it depending on Python for anything).

You might still be able to receive a PyDict of type `{ node_type: duration_type }` while keeping a path open for rust users to use the respective rust native types. Which would allow you to remove the python types altogether for a rust side without changing anything from what you've done on the Python side.